### PR TITLE
`Development`: Add architecture test for `@Param` usage

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/

### Motivation and Context
The Artemis server style guide asks for the `@Param` annotation to always be used when creating a `@Query` repository method.

### Description
This PR adds an architecture test to ensure the annotation gets added.

### Steps for Testing
code review

#### Code Review
- [x] Code Review 1
- [x] Code Review 2

## Summary by CodeRabbit

- **Refactor**
	- Updated method signatures across various repositories to improve parameter handling and query efficiency.
	- Simplified method signatures by removing or adding `@Param` annotations where necessary for clearer and more efficient database queries.
	- Enhanced consistency in the use of JPQL query keywords and parameter annotations across the application.

- **Chores**
	- Added new import statements for `org.springframework.data.repository.query.Param` in several repository files to support the refactor changes.

- **Tests**
	- Introduced new test methods in `ArchitectureTest.java` to ensure the correct use of `@Param` annotations and to verify the absence of direct Git commit calls in repository methods.

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/ArchitectureTest.java [hunk 1]
```diff
@@ -160,14 +181,24 @@ public void check(JavaMethod item, ConditionEvents events) {
...
-    /**
-     * Checks that no class directly calls Git.commit(), but instead uses GitService.commit()
-     * This is necessary to ensure that committing is identical for all setups, with and without commit signing
-     */
-    @Test
-    void testNoDirectGitCommitCalls() {
-        ArchRule usage = noClasses().should().callMethod(Git.class, "commit").because("You should use GitService.commit() instead");
-        var classesWithoutGitService = allClasses.that(not(assignableTo(GitService.class)));
-        usage.check(classesWithoutGitService);
+    private ArchCondition<JavaMethod> haveAllParametersAnnotatedWith(DescribedPredicate<? super JavaAnnotation<?>> annotationPredicate, DescribedPredicate<JavaClass> exception) {
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/ArchitectureTest.java:184]: Can you please imrove the name? Right now it reads, without checking the signature, that they can be annotated with one of them.
Maybe `haveAllParametersAnnoatedWithUnless` or something like that?

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java [hunk 2]
```diff
@@ -36,7 +36,7 @@
 @Repository
 public interface StudentParticipationRepository extends JpaRepository<StudentParticipation, Long> {
 
-    Set<StudentParticipation> findByExerciseId(@Param("exerciseId") Long exerciseId);
+    Set<StudentParticipation> findByExerciseId(Long exerciseId);
 
     @Query("""
             SELECT DISTINCT p
```
[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:42]: Why would `JOIN FETCH` improve the performance compared to `LEFT JOIN FETCH`?

**Review by rstief — APPROVED** [review]
Code looks good 👍

File: src/test/java/de/tum/in/www1/artemis/ArchitectureTest.java [hunk 3]
```diff
@@ -160,14 +183,22 @@ public void check(JavaMethod item, ConditionEvents events) {
...
-        usage.check(classesWithoutGitService);
+    private ArchCondition<JavaMethod> haveAllParametersAnnotatedWithUnless(DescribedPredicate<? super JavaAnnotation<?>> annotationPredicate,
+            DescribedPredicate<JavaClass> exception) {
+        return new ArchCondition<>("have all parameters annotated with " + annotationPredicate.getDescription()) {
+
+            @Override
+            public void check(JavaMethod item, ConditionEvents events) {
+                boolean satisfied = item.getParameters().stream()
+                        // Ignore annotations of the Pageable parameter
+                        .filter(javaParameter -> !exception.test(javaParameter.getRawType())).map(JavaParameter::getAnnotations)
```
Comment (rstief) [inline src/test/java/de/tum/in/www1/artemis/ArchitectureTest.java:194]: I don't quite understand how getRawType() (only) filters out the pageable parameters, can you elaborate a bit on that?

**Review by rstief — APPROVED** [review]

**Review by b-fein — APPROVED** [review]

**Pull request merged**